[python-package] Create Dataset from multiple data files

cyfdecyf opened the pull request:
This PR implements creating Dataset from multiple data files.

We add a `Sequence` interface which requires data providers implement:

- Return total number of data
- Random access
  - so we can sample data without going through data in two round
- Range access
  - used to read data in batch thus avoid loading all data into memory

With this interface, we can create `Dataset` from `List[Sequence]`. Also refer this to #2789 as I think this interface allows we use other types of efficient binary data format. I closed #3900 because it's an ugly hack and data loading features are much easier to implement in Python.

An example of creating `Dataset` from multiple HDF5 files is also given. As I mentioned in #2788, I'm more familiar with HDF5 file so it's choosen as the example.

 #2031 also mentions the feature request, guess this is a common feature request.

cyfdecyf commented on python-package/lightgbm/basic.py:
```diff
@@ -594,6 +600,61 @@ def _load_pandas_categorical(file_name=None, model_str=None):
+        # Random access by row index. Use for data sampling.
+        >>> seq[10]
+        # Range data access. Use to read data in batch when constructing Dataset.
+        >>> seq[0:100]
```
I'm also considering use a separate interface such as `next_batch` to iterate over data in batch. So we don't need range access and there's no need to set batch size.

Willian-Zhang replied:
comparison between current implementation sequence and `next`:

### Current
- LightGBM determines how to read, in-order out-of-order etc
- Data Provides hints reading behavior via `batch_size`
##### Potential improvements
- generally prefetching, ultimate form be like [auto-data-tunning ](https://www.tensorflow.org/guide/data_performance) in TensorFlow 
- parallel reading, even distributed reading
- natural to go with MPI (MPI I/O) Interface 

### `next`
- simplicity 
- data provider get to control what and how to read
- data provider could optimize data feeding accordingly (no need to wait for two segment of data from different cassette be fully loaded, just early cut when one is loaded and hint an prefetch event) 
##### future
- total separation of sampling and data reading stages, each could benefits in someway
- infinite dataset support 


To compare:
- both changes to current implementation and `next` approach could improve efficiency upon current implemntation
- current being more general, could leverage speed for almost any data source once sophisticated prefetch mechanism are implemented
- `next` being simplistic, giving all efficiency rights to data provider (note, data providers are still not limited in current implementation)

Willian-Zhang replied:
just to note, these two implement do not conflict each other, in fact, to their ultimate goal, they could well co-exist to compensate each other.

shiyu1994 replied:
Is current approach more convenient and general for many potential accelerations? For prefetching, I think both current approach and the batch approach can do. But for parallel reading, it seems current approach is much more convenient, unless the data provider implements the `next` method with little overhead when a batch is not actually accessed. The current approach seems to put more burden on the data provider implementation though.

For prefetching when reading full dataset, I think we can pipeline reading from the `seq` and `push_rows` here.
https://github.com/cyfdecyf/LightGBM/blob/1c2a6ef185e813b11a1601764de09d5d576503cc/python-package/lightgbm/basic.py#L1575-L1580
If reading from the `seq` is I/O bounded, pipelining it with `push_rows` may hide the I/O latency.

Willian-Zhang replied:
> Is current approach more convenient and general for many potential accelerations?
While I do think current approach is more general, it limits how data provider could optimize their own pipeline, since there is no order guarantee for them. On the other hand, LightGBM, since have the full control, can be optimized, like pipelining as you mentioned.

Idea here is, 
- If we trust user could build the best data provider against their own data storage traits, we should put limitations on the interface, so that data provider could get an garenteed data access pattern. This limits what lightGBM could do, yet maximizes global potential.
- If we don't, maybe we should keep using current general implementation, that we could do more (pipelining, prefetching, multiprocessing, distributed I/O, auto-tunnning, etc etc) and benefits everyone that implements it.
- I would say those two both has their benefits, if the user base is large enough, I would recommend implementing both. - (or even consider separate data sampling stage and parsing stage, allow the other one to be implemented in a total different way)

> For prefetching when reading full dataset, I think we can pipeline reading from the seq and push_rows here.

That seems like a very good idea. Are you implying multithreading? 

Large data reading could bounded by CPU (as long as they are not implemented in hardware offloaded fashion e.g. MPI), I'm not so sure if that would be the best.

Willian-Zhang commented:
Other potential improvements:
`Sequence` objects can be easily extended with support of batch reading with row numbers:
``` python
seq[[1,2,5,7,9,....]]
```
such, data provider could provide more efficient reading than current one-row-at-time fashion under the hood.

Willian-Zhang commented on python-package/lightgbm/basic.py:
```diff
@@ -1383,8 +1439,15 @@ def _lazy_init(self, data, label=None, reference=None,
             self.__init_from_np2d(data, params_str, ref_dataset)
-        elif isinstance(data, list) and len(data) > 0 and all(isinstance(x, np.ndarray) for x in data):
-            self.__init_from_list_np2d(data, params_str, ref_dataset)
+        elif isinstance(data, Sequence):
```
TODO: revert `is_class` implementation 

This would force any 3rd party lib (presumably Data Access Layer libs) to import LightGBM.

Willian-Zhang replied:
Generally we need an duck typing implementation.
 `typechecked` and `is_of_type` would do the job yet requires new dependency [pytypes](https://pypi.org/project/pytypes/).

Willian-Zhang reviewed (changes requested):
revert is_class implementation or use pytypes

Willian-Zhang commented on python-package/lightgbm/basic.py:
```diff
@@ -1297,6 +1475,74 @@ def _lazy_init(self, data, label=None, reference=None,
+                seq = seqs[seq_id]
+            id_in_seq = row_id - offset
+            row = seq[int(id_in_seq)]
+            yield row
```
TODO: fix unwanted memory view reference

Willian-Zhang replied:
Either user or LGBM should be doing for once.
Are LGBM users generally trusted in design philosophy?

Willian-Zhang reviewed (changes requested):
TODO: fix unwanted memory reference

Willian-Zhang reviewed (approved):
LGTM

Willian-Zhang commented:
@shiyu1994 this PR is now ready. Are you still available for reviewing?

shiyu1994 commented on python-package/lightgbm/basic.py:
```diff
@@ -1297,6 +1492,74 @@ def _lazy_init(self, data, label=None, reference=None,
+
+        return filtered, filtered_idx
+
+    def __init_from_seqs(self, seqs, params_str, ref_dataset):
```
Unused parameters `params_str` and `ref_dataset`. It seems that `ref_dataset` is ignored when we constructing from `seqs`. Does that mean this method only supports constructing training dataset? Because for a validation dataset, a reference training dataset is required in Python API.

Willian-Zhang replied:
Very much appreciated for making clear on what `ref_dataset` does.
 
>  Does that mean this method only supports constructing training dataset? 
 
Currently, yes. However after your explanation, It seems validation dataset can also be supported via trivial changes to make.

shiyu1994 replied:
Ok. I think now that we've supported creating from `seqs` for training data, we should also support that for validation data. Otherwise, users can get unexpected behavior if they try to create validation dataset from `seqs`. Because the only difference between validation and training datasets in Python API is whether `reference` parameter in `lightgbm.Dataset` is provided.

I'm willing to provide any help.

shiyu1994 commented on python-package/lightgbm/basic.py:
```diff
@@ -1297,6 +1492,74 @@ def _lazy_init(self, data, label=None, reference=None,
+            Supports random access and access by batch if properly defined by user
+        """
+        total_nrow = sum(len(seq) for seq in seqs)
+        ncol = len(seqs[0][0])
```
Unused variable `ncol`.

Willian-Zhang replied:
This was left out for debugging propose and really should be further extended for data sanity check before loading them.

Willian-Zhang replied:
However data check could impact performance since `Sequence` interface keeps data access unknown to LightGBM.

I suggest adding a new toggle parameter for checking to control such behavior.

shiyu1994 replied:
Do you mean we need to check that each `seq` has exactly the same number of columns? Doing that here is perfectly ok and won't incur efficiency degradation.

Willian-Zhang replied:
> Do you mean we need to check that each `seq` has exactly the same number of columns? 

exactly

> Doing that here is perfectly ok and won't incur efficiency degradation.

Actually we wouldn't know that for sure. 
To exemplify, say our user decided to implement some `seq`s with some very heavy data provider backend and somehow implemented data prefetching and parallelism as their optimization. (In extreme cases imagine data are generated on the fly from some server instances spawned upon our `touching` to the data), 

They might would trigger the generation/read for their whole `seq` when we touch the `seq` since they would expect data are accessed in ordered fashion.

To generalize, we really cannot make assumptions on how much a single access to data would cost. we **hope** they would provide best performance for both **random data access** and **by batch access**.  However in reality, these two are naturally conflicting each other in someway. Assuring batch access is in-order, at least giving an option to do so, is the best we can do to that.

Willian-Zhang replied:
> I suggest adding a new parameter for toggling checking behavior.

Actually some helps to this would be very much helpful.

Is it possible to create a `param` for python only, in particular, not to pass down to or cause cpp to warn unrecognized parameter?

shiyu1994 replied:
Yes, that is possible. We need to pick out the parameter in Python before passing it to CPP, which is not an elegant solution. 
Also, we'd better keep the set of parameters in all API consistent.

Willian-Zhang replied:
fixed via https://github.com/microsoft/LightGBM/pull/4089/commits/87ef72f1c006a7434f687fd404cd926b86167cbc

shiyu1994 commented on src/io/dataset_loader.cpp:
```diff
@@ -974,6 +1002,22 @@ void DatasetLoader::ConstructBinMappersFromTextData(int rank, int num_machines,
+
+  /*
+  int num_col = static_cast<int>(sample_values.size());
+  std::vector<double*> sample_values_ptr(num_col);
```
Useless code should be deleted.

cyfdecyf replied:
This is also for debugging. If you like the idea of keep debugging code, I'd move those code into a separarte function that would only be called when debugging is enabled.

shiyu1994 commented on src/io/dataset_loader.cpp:
```diff
@@ -615,10 +615,38 @@ Dataset* DatasetLoader::LoadFromBinFile(const char* data_filename, const char* b
 }
 
+// To help verify whether sample data is changed when using different language bindings.
+static void DumpSampleData(const std::string& sample_filename, double** sample_values,
```
I'm not sure whether this method is necessary to be included in the master branch. Is it for debugging? It seems that all calls to `ConstructFromSampleData` in the project will pass an empty `dump_filename`, so this function is not called in the source code actually. And no API is provided for users to access this method.

cyfdecyf replied:
Yes, this is for debugging. I suggest we find some acceptable way to keep these debugging code. It helped a lot to find problems when we develop this feature.

cyfdecyf replied:
BTW, c api exposes the function `LGBM_DatasetDumpText` for debugging. How about we add an option (say `dump_sample`) when creating Dataset in Python API?

shiyu1994 replied:
I think we'd better remove these debug code, which will increase maintenance effort. Unlike dumping a whole `Dataset`, dumping the sampling part used by construction is seldom needed. And I believe we can manually do so whenever necessary in development.
Thanks again for your contribution.

cyfdecyf replied:
OK. I'll remove those debugging code this week.

cyfdecyf replied:
@shiyu1994 Debugging code removed now. Please let me know if there's anything that should be improved.

shiyu1994 commented on tests/python_package_test/test_basic.py:
```diff
@@ -84,6 +93,88 @@ def test_basic(tmp_path):
+@pytest.mark.parametrize('include_nan', [False, True])
+@pytest.mark.parametrize('num_seq', [1, 3])
+def test_sequence(tmpdir, sample_count, batch_size, include_0, include_nan, num_seq):
+    rm_files(["seq.truth.bin", "seq.seq.bin"])
```
It seems that `rm_files` removes `seq.truth.bin` and `seq.seq.bin`, but subsequently `tmpdir/seq.truth.bin` and `tmpdir/seq.seq.bin` will be created. That's a mismatch.
And why do we need to manually remove the files created under `tmpdir` here, which is not done in other testing cases?

Willian-Zhang replied:
> why do we need to manually remove the files created under tmpdir here, which is not done in other testing cases?
we don't. tmpdir is unique per parameter.

This is indeed a left-out from previous code where `tmpdir`  was not used.

Will fix.

shiyu1994 commented:
@cyfdecyf Thanks for your great contribution! I'm comment about my concerns. Overall the implementation is great, but there's seems to be some unnecessary changes.

Willian-Zhang commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1193,123 @@ def __init__(self, data, label=None, reference=None,
+        return indices
+
+    def init_from_sample(self, sample_data, sample_indices, sample_cnt, total_nrow):
+        """Get the used parameters in the Dataset.
```
Do you mean `Get the used parameters in the Dataset.` wrongfully describes what this method does?

Yes, I confirm we lost track of where this summary was from. Nevertheless, will fix.

shiyu1994 commented:
@cyfdecyf Thanks for removing the debugging code. Would you like to continue to complete this PR for the validation datasets? If not, I can help to do that.

Willian-Zhang commented:
> @cyfdecyf Thanks for removing the debugging code. Would you like to continue to complete this PR for the validation datasets? If not, I can help to do that.

It's on plan, will do in 10 days.

jameslamb commented on .ci/test.sh:
```diff
@@ -57,9 +57,13 @@ if [[ $TASK == "lint" ]]; then
     pip install --user cpplint isort mypy
     echo "Linting Python code"
+    echo "..pycodestyle"
     pycodestyle --ignore=E501,W503 --exclude=./.nuget,./external_libs . || exit -1
-    pydocstyle --convention=numpy --add-ignore=D105 --match-dir="^(?!^external_libs|test|example).*" --match="(?!^test_|setup).*\.py" . || exit -1
+    echo "..pydocstyle"
+    pydocstyle --convention=numpy --add-ignore=D105 --match-dir='^(?!^external_libs|test|example).*' --match='(?!^test_|setup).*\.py' . || exit -1
+    echo "..isort"
     isort . --check-only || exit -1
+    echo "..mypy"
```
Can you please remove these changes adding new `echo` statements and moving the order of CI checks around? They don't seem related to this PR.

Willian-Zhang replied:
fixed via 941dfce43418523067a3edf5bb93a1883fee38a2

jameslamb commented on src/io/dataset.cpp:
```diff
@@ -927,7 +927,7 @@ bool Dataset::GetIntField(const char* field_name, data_size_t* out_len,
 void Dataset::SaveBinaryFile(const char* bin_filename) {
   if (bin_filename != nullptr && std::string(bin_filename) == data_filename_) {
-    Log::Warning("Bianry file %s already exists", bin_filename);
+    Log::Warning("Binary file %s already exists", bin_filename);
```
Thanks for fixing this! But please remove this unrelated change and make a separate pull request for it.

In general, we have a preference for many small pull requests over a few large ones. A pull request that just changes this typo would be easy to review and merged very quickly, and then the size of the diff for this PR would be reduced.

cyfdecyf replied:
OK. I'll removing this first.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -594,6 +599,79 @@ def _load_pandas_categorical(file_name=None, model_str=None):
+        raise NotImplementedError("remove this line if subclassing")
+
+    @abc.abstractmethod
+    def __len__(self):  # type: () -> int
```
```suggestion
    def __len__(self) -> int:
```

Is there a reason that you prefer to use type hint comments instead of type hints in code? As of https://github.com/microsoft/LightGBM/pull/3581 this project no longer supports Python 2.x, so you can use type hints in code directly.

In fact, based on #3756 we have a preference for it.

jameslamb replied:
Unless there is a specific reason to use comments, can you please remove them in this PR and replace them with type hints in code? For consistency with the rest of the project.

cyfdecyf replied:
It's for Python 2.x support. It's a pleasure to drop Python 2.x support.

Willian-Zhang replied:
Will do +1

jameslamb reviewed (changes requested):
Thanks for this, @cyfdecyf ! I left some comments related to consistency with LightGBM's existing codebase and removing unrelated changes from this PR.

I haven't reviewed the actual functional changes in the PR yet, since it seems like there is still some ongoing discussion about it (e.g. https://github.com/microsoft/LightGBM/pull/4089/files#r598454555, https://github.com/microsoft/LightGBM/pull/4089/files#r612925422, https://github.com/microsoft/LightGBM/pull/4089/files#r612950521)

cyfdecyf commented on python-package/lightgbm/basic.py:
```diff
@@ -594,6 +599,79 @@ def _load_pandas_categorical(file_name=None, model_str=None):
+    reduce memory usage.
+    """
+
+    __metaclass__ = abc.ABCMeta
```
In `Dataset._lazy_init` there's many `isinstance` calls to decide which init function to use. Using `abc` allows us to keep use `isinstance` and keep consistency there.

The initial PR does not contain `is_class` method and thus this is necessary. For now, using `abc` is not necessary and we can just remove it.

Willian-Zhang replied:
Just to add some more context:

> Generally we need an duck typing implementation.
 `typechecked` and `is_of_type` would do the job yet requires new dependency [pytypes](https://pypi.org/project/pytypes/).

_Originally posted by @Willian-Zhang in https://github.com/microsoft/LightGBM/pull/4089#discussion_r602071148_


`pytypes` was originally being used for duck typing implementation. 
It is removed in consider to minimize new dependency added to LightGBM.
`is_class` as along some reordering on `is_instance`s  was added as a trade-off.
@cyfdecyf has some good reason to adopt `ABC` route as stated in this thread, and here we are.

Currently, LightGBM tries to call `is_class` from this commit:
https://github.com/microsoft/LightGBM/pull/4089/commits/a8bc7d99eab660db620dfdbdb525e0650c7ca79b#diff-5e0fb2a7b1ec4988ccde5e74a6c4b609841c1143f281da2afc85e8c464d37b3bR625-R630
to determine if a class instance ducks `Sequence` dues to ABC implementation would result in LightGBM misinterpreting non-subclassing `seq` object to other types. (convertible to ndarray or csr_matrix, cannot recall exact one.)

Willian-Zhang replied:
fixed via https://github.com/microsoft/LightGBM/pull/4089/commits/591fc5e6b1902346f7e3ae6b8b38ad2aa4cefb1d

Willian-Zhang commented on tests/python_package_test/test_basic.py:
```diff
@@ -84,6 +93,86 @@ def test_basic(tmp_path):
+            seqs.append(seq)
+    ds = lgb.Dataset(seqs, label=Y, params=params)
+    ds.save_binary(str(tmpdir / "seq.seq.bin"))
+    assert filecmp.cmp(tmpdir / "seq.truth.bin", tmpdir / "seq.seq.bin")
```
Will do

Willian-Zhang commented:
> @cyfdecyf Thanks for removing the debugging code. Would you like to continue to complete this PR for the validation datasets? If not, I can help to do that.

validation dataset is completed with via https://github.com/microsoft/LightGBM/pull/4089/commits/92d0b39cbf357f482638e06e57e184a1c8f52aed

tho it seems saved binary file for validation datasets from same data are not guaranteed to stay identical. 
I'm not so sure if this is an expected behavior. 


Some observations:
![image](https://user-images.githubusercontent.com/5165840/117271661-fee4e400-ae8c-11eb-937b-79fbac3e49a2.png)
```
[LightGBM] [Info] num_features_ 0
[LightGBM] [Info] num_total_features_ 10
[LightGBM] [Info] label_idx_ 0
[LightGBM] [Info] max_bin_ 0
[LightGBM] [Info] bin_construct_sample_cnt_ -1359898024
[LightGBM] [Info] min_data_in_bin_ 32723
[LightGBM] [Info] zero_as_missing_ 248
[LightGBM] [Info] has_raw_ 0
```

shiyu1994 commented on src/c_api.cpp:
```diff
@@ -1034,6 +1034,35 @@ int LGBM_DatasetCreateFromMat(const void* data,
+
+  auto sample_indices = CreateSampleIndices(config, total_nrow);
+
+  static_assert (sizeof(int) == 4, "int size is not 4");
```
What's the purpose of this assert statement?

Willian-Zhang replied:
@cyfdecyf

cyfdecyf replied:
This is useless now and removed.

`CreateSampleIndicies` returns `std::vector<int>` at first. It's now returning `std::vector<int32_t>` by implicit type convertion from `Random.Sample`'s return value.

shiyu1994 commented on tests/python_package_test/test_basic.py:
```diff
@@ -13,6 +14,14 @@
+        files = [files]
+    for file in files:
+        if os.path.exists(file):
+            os.remove(file)
```
As mentioned before, this should be removed.

Willian-Zhang replied:
removed via https://github.com/microsoft/LightGBM/pull/4089/commits/2d32b873db3ab7b436e2941389eceb9817290d8d

shiyu1994 commented on tests/python_package_test/test_basic.py:
```diff
@@ -84,6 +93,109 @@ def test_basic(tmp_path):
+    ds.save_binary(os.path.join(tmpdir, "seq.seq.bin"))
+
+    if create_valid:
+        # TODO: verify validation dataset somehow
```
Add tests for validation dataset.

shiyu1994 commented:
> ```
> [LightGBM] [Info] num_features_ 0
> [LightGBM] [Info] num_total_features_ 10
> [LightGBM] [Info] label_idx_ 0
> [LightGBM] [Info] max_bin_ 0
> [LightGBM] [Info] bin_construct_sample_cnt_ -1359898024
> [LightGBM] [Info] min_data_in_bin_ 32723
> [LightGBM] [Info] zero_as_missing_ 248
> [LightGBM] [Info] has_raw_ 0
> ```
Are these attributes from a validation dataset loaded from bin file?
For validation dataset, `num_features_ 0` is not expected, since it should be set in `Dataset::CreateValid`, which is called by `LGBM_DatasetCreateByReference`.

https://github.com/microsoft/LightGBM/blob/f83180883a51729af9562ccac7b450c6da79a04a/src/io/dataset.cpp#L726-L728

shiyu1994 commented:
Any progress on the validation data issue? Maybe I can help to debug this.

Willian-Zhang commented:
that would be very helpful

cyfdecyf commented:
@shiyu1994 @Willian-Zhang

#4177 fixed valid dataset with uninitialized instance variables.

`num_features_ 0` seems like caused by too small sample count count. After increasing smallest sample count to 11, this problem is also solved.

shiyu1994 commented on src/io/dataset.cpp:
```diff
@@ -932,7 +932,7 @@ bool Dataset::GetIntField(const char* field_name, data_size_t* out_len,
 void Dataset::SaveBinaryFile(const char* bin_filename) {
   if (bin_filename != nullptr && std::string(bin_filename) == data_filename_) {
-    Log::Warning("Binary file %s already exists", bin_filename);
+    Log::Warning("Bianry file %s already exists", bin_filename);
```
It seems that this reverts the typo fixing.

cyfdecyf replied:
Sigh... thanks for spoting this! Fix pushed.

shiyu1994 commented:
@cyfdecyf Thanks a lot. We are almost ready, except for a typo issue.

shiyu1994 commented:
@jameslamb Could you please check if your change requests have been addressed?

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1188,144 @@ def __init__(self, data, label=None, reference=None,
+        ))
+        return indices
+
+    def init_from_ref_dataset(self, total_nrow: int, ref_dataset: 'Dataset'):
```
```suggestion
    def init_from_ref_dataset(self, total_nrow: int, ref_dataset: 'Dataset') -> 'Dataset':
```

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1188,144 @@ def __init__(self, data, label=None, reference=None,
+        ))
+        return self
+
+    def init_from_sample(self, sample_data, sample_indices, sample_cnt, total_nrow):
```
```suggestion
    def init_from_sample(
        self,
        sample_data: List[np.array[float64]],
        sample_indices: List[np.array[int]],
        sample_cnt: int,
        total_nrow: int
    ) -> "Dataset":
```

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1188,144 @@ def __init__(self, data, label=None, reference=None,
+        self : Dataset
+            Constructed Dataset object.
+        """
+        assert len(sample_data) == len(sample_indices), "#sample data column != #column indices"
+
+        ncol = len(sample_indices)
```
```suggestion
        ncol = len(sample_indices)
        assert len(sample_data) == ncol, "#sample data column != #column indices"

```

These could be arranged to avoid computing `len(sample_indices)` twice.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1188,144 @@ def __init__(self, data, label=None, reference=None,
+        ))
+        return self
+
+    def push_rows(self, data):
```
```suggestion
    def push_rows(self, data: np.ndarray) -> 'Dataset':
```

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1297,6 +1508,77 @@ def _lazy_init(self, data, label=None, reference=None,
+        if ref_dataset:
+            self.init_from_ref_dataset(total_nrow, ref_dataset)
+        else:
+            sample_cnt = self.params.get("bin_construct_sample_cnt") or DEFAULT_BIN_CONSTRUCT_SAMPLE_CNT
```
```suggestion
            sample_cnt = self.params.get("bin_construct_sample_cnt", DEFAULT_BIN_CONSTRUCT_SAMPLE_CNT)
```

jameslamb commented on src/c_api.cpp:
```diff
@@ -1035,6 +1035,33 @@ int LGBM_DatasetCreateFromMat(const void* data,
+}
+
+
+int LGBM_SampleIndices(int32_t total_nrow,
+                const char* parameters,
+                void* out) {
```
```suggestion
int LGBM_SampleIndices(int32_t total_nrow,
                       const char* parameters,
                       void* out) {
```

Please align these arguments vertically, similar to the style used in the rest of this file. This makes it a bit easier to read through the list of parameters.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -594,6 +598,75 @@ def _load_pandas_categorical(file_name=None, model_str=None):
         return None
 
 
+class Sequence(object):
```
```suggestion
class Sequence:
```

Since LightGBM no longer supports Python 2 (https://github.com/microsoft/LightGBM/pull/4089#discussion_r620355287), it isn't necessary to explicitly declare that a class inherits from `object`. We have a preference for not using `(object)` in class declarations. For example, see these other class declarations in this file:

https://github.com/microsoft/LightGBM/blob/32fec820607d8cbcd1fe37c200455ea45702df77/python-package/lightgbm/basic.py#L1047

https://github.com/microsoft/LightGBM/blob/32fec820607d8cbcd1fe37c200455ea45702df77/python-package/lightgbm/basic.py#L2150

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1188,144 @@ def __init__(self, data, label=None, reference=None,
             pass
 
+    def create_sample_indices(self, total_nrow):
+        """Create sample indices for the given parameter of the Dataset.
```
```suggestion
        """Get an array of randomly chosen indices from this ``Dataset``.
        
        Indices are sampled without replacement.
```

I was confused by this description. Would you consider this change? Hopefully I've understood this method correctly.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1188,144 @@ def __init__(self, data, label=None, reference=None,
+        ----------
+        total_nrow: int
+            Total number of rows to sample from.
+            If Dataset has multiple input data, this should be the sum of rows of every file.
```
```suggestion
            If this value is greater than the value of parameter ``bin_construct_sample_cnt``, only ``bin_construct_sample_cnt`` indices will be used.
            If Dataset has multiple input data, this should be the sum of rows of every file.
```

My first thought reading this was "what would happen if the training data has `total_nrow` is larger than the number of samples?". Looking at the changes, I think I understand the behavior. If I've understood correctly, would you please consider this documentation change?

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1188,144 @@ def __init__(self, data, label=None, reference=None,
+    def init_from_ref_dataset(self, total_nrow: int, ref_dataset: 'Dataset'):
+        """Create dataset from a reference dataset.
+
+        Args
+        ----------
```
```suggestion
        Parameters
        ---------------
```

This project's documentation style uses `Parameters`, not `Args` for the list of parameters. You can look at the other docstrings in this file for reference.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1188,144 @@ def __init__(self, data, label=None, reference=None,
+            total_nrow (int): number of rows expected to add to dataset
+            ref_dataset (Dataset): referance dataset to extract meta from
+
+        Returns
+        -------
+            self : Dataset
+                Constructed Dataset object.
```
```suggestion
        Returns
        -------
        self : Dataset
            Constructed Dataset object.
```

The return type should be vertically aligned with its heading. See the other docstrings in this file for examples.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1115,13 +1188,144 @@ def __init__(self, data, label=None, reference=None,
+    def push_rows(self, data):
+        """Add rows to Dataset.
+
+        Args:
+            data: numpy 1-D array
```
```suggestion
        Parameters
        ----------
        data : numpy array
            New data to add to the Dataset.
```

jameslamb reviewed (changes requested):
Thanks very much for this! Please see my most recent (minor) comments. For any that have a suggestion, you can click "add to batch" or "commit suggestion(s)" to commit them directly in the browser.

cyfdecyf commented:
@jameslamb Thanks for the detailed review! I've left two review comments and will will resolve them in a few days.

cyfdecyf commented:
@jameslamb It's weird that there's lot's of CI failures. I don't find a way to run those failing tests locally.

So I rebased to the latest master and reset my branch to the commit that has CI passed to see if the problem is caused by my later commits.

I'll comment when I've finished. Sorry for the force push noise in this issue.

cyfdecyf commented:
@jameslamb Resolved all recent review comments and fixed CI.

Note that type hinting like following can't pass lint check.

```
def init_from_sample(
        self,
        sample_data: List[np.ndarray[np.float64]],
        sample_indices: List[np.ndarray[int]],
```

Seems like we need use the following:

```
import numpy.typing as npt

def init_from_sample(
        self,
        sample_data: List[npt.NDArray[np.float64]],
        sample_indices: List[np.NDArray[np.int]],
```

I'm not very family with Python type hinting. As [numpy document](https://numpy.org/doc/stable/reference/typing.html) says some features require Python 3.8, so I just changed type hint to `List[np.ndarray]`.

jameslamb commented:
> @jameslamb Resolved all recent review comments and fixed CI.
> 
> Note that type hinting like following can't pass lint check.
> 
> ```
> def init_from_sample(
>         self,
>         sample_data: List[np.ndarray[np.float64]],
>         sample_indices: List[np.ndarray[int]],
> ```
> 
> Seems like we need use the following:
> 
> ```
> import numpy.typing as npt
> 
> def init_from_sample(
>         self,
>         sample_data: List[npt.NDArray[np.float64]],
>         sample_indices: List[npt.NDArray[np.int]],
> ```
> 
> I'm not very family with Python type hinting. As [numpy document](https://numpy.org/doc/stable/reference/typing.html) says some features require Python 3.8, so I just changed type hint to `List[np.ndarray]`.

thanks very much for the investigation. I think `List[np.ndarray]` is ok for now.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -9,14 +9,18 @@
 from .compat import PANDAS_INSTALLED, concat, dt_DataTable, is_dtype_sparse, pd_DataFrame, pd_Series
 from .libpath import find_lib_path
 
+# TODO: how to keep the default values the same with C++ config.h
```
closing this since #4321 was created, thank you!

jameslamb commented on .ci/test_windows.ps1:
```diff
@@ -49,7 +49,7 @@ if ($env:TASK -eq "swig") {
 }
 
-conda install -q -y -n $env:CONDA_ENV joblib matplotlib numpy pandas psutil pytest python-graphviz scikit-learn scipy ; Check-Output $?
+conda install -q -y -n $env:CONDA_ENV joblib matplotlib numpy pandas psutil pytest python-graphviz scikit-learn scipy h5py ; Check-Output $?
```
looks like this has been addressed, thanks very much

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1183,155 @@ def __init__(self, data, label=None, reference=None,
+
+        Parameters
+        ----------
+        sample_data:
+            Sample data for each column
+        sample_indices:
+            Sample data row index for each column.
+        sample_cnt:
+            Number of samples.
+        total_nrow:
+            Total number of rows for all input file.
```
```suggestion
        sample_data : list of numpy arrays
            Sample data for each column
        sample_indices : list of numpy arrays
            Sample data row index for each column.
        sample_cnt : int
            Number of samples.
        total_nrow : int
            Total number of rows for all input file.
```

Can you please add type hints in these parameter docs? So they will show up in the rendered docs.

This is what the docs for `Dataset.init_from_sample()` currently look like without those:

![image](https://user-images.githubusercontent.com/7608904/120585033-a9054c80-c3f6-11eb-9223-af061ad1c598.png)

Compare that to a method with hints

![image](https://user-images.githubusercontent.com/7608904/120585083-bc181c80-c3f6-11eb-920a-08f72df7227d.png)

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1183,155 @@ def __init__(self, data, label=None, reference=None,
+
+        for i in range(ncol):
+            if sample_data[i].dtype != np.double:
+                raise ValueError("sample_data[{}] type {} is not double".format(i, sample_data[i].dtype))
```
As of #4143, as part of #4136, all string formatting in this file has been converted to use f-strings, for speed and ease of use.

Can you please change this and any other uses of `.format()` to f-strings?

```suggestion
                raise ValueError(f"sample_data[{i}] type {sample_data[i].dtype} is not double")
```

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1183,155 @@ def __init__(self, data, label=None, reference=None,
+            if sample_data[i].dtype != np.double:
+                raise ValueError("sample_data[{}] type {} is not double".format(i, sample_data[i].dtype))
+            if sample_indices[i].dtype != np.int32:
+                raise ValueError("sample_indices[{}] type {} is not int32".format(i, sample_indices[i].dtype))
```
```suggestion
                raise ValueError(f"sample_indices[{i}] type {sample_indices[i].dtype} is not int32")
```

jameslamb commented on tests/python_package_test/test_basic.py:
```diff
@@ -89,6 +90,115 @@ def test_basic(tmp_path):
+                raise NotImplementedError("No need to implement, caller will not set step by now")
+            return self.ndarray[idx.start: idx.stop]
+        else:
+            raise TypeError("Sequence Index must be an integer/list/slice, got {}".format(type(idx)))
```
```suggestion
            raise TypeError(f"Sequence Index must be an integer/list/slice, got {type(idx)}")
```

jameslamb commented on examples/python-guide/dataset_from_multi_hdf5.py:
```diff
@@ -0,0 +1,102 @@
+
+    # We can store multiple datasets inside a single HDF5 file.
+    # Separating X and Y for choosing best chunk size for data loading.
+    save2hdf({'Y': df1.iloc[:, :1], 'X': df1.iloc[:, 1:]}, '{}1.h5'.format(output_basename), batch_size)
+    save2hdf({'Y': df2.iloc[:, :1], 'X': df2.iloc[:, 1:]}, '{}2.h5'.format(output_basename), batch_size)
```
```suggestion
    save2hdf({'Y': df1.iloc[:, :1], 'X': df1.iloc[:, 1:]}, f'{output_basename}1.h5', batch_size)
    save2hdf({'Y': df2.iloc[:, :1], 'X': df2.iloc[:, 1:]}, f'{output_basename}2.h5', batch_size)
```

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -591,6 +594,75 @@ def _load_pandas_categorical(file_name=None, model_str=None):
+            elif isinstance(idx, slice):
+                return np.stack(self.__get_one_line__(i) for i in range(idx.start, idx.stop))
+            else:
+                raise TypeError("Sequence index must be integer or slice, got {}".format(type(idx)))
```
```suggestion
                raise TypeError(f"Sequence index must be integer or slice, got {type(idx)}")
```

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -591,6 +594,75 @@ def _load_pandas_categorical(file_name=None, model_str=None):
+        result : numpy 1-D array, numpy 2-D array
+            1-D array if idx is int, 2-D array if idx is slice
+        """
+        raise NotImplementedError("remove this line if subclassing")
```
```suggestion
        raise NotImplementedError("Sub-classes of lightgbm.Sequence must implement __getitem__()")
```

Could this be made a bit more specific? I have worked on systems before where error messages are caught and logged, without the accompanying stack trace, so messages like "this line" can be difficult or impossible to trace back to the source.

jameslamb commented on python-package/lightgbm/basic.py:
```diff
@@ -591,6 +594,75 @@ def _load_pandas_categorical(file_name=None, model_str=None):
+
+    def __len__(self) -> int:
+        """Return row count of this sequence."""
+        raise NotImplementedError
```
```suggestion
        raise NotImplementedError("Sub-classes of lightgbm.Sequence must implement __len__()")
```

Could this have a more specific error message, similar to the one I just suggested for `__getitem__()`? I think that will help debugging cases where someone uses code that subclassed `Sequence` but failed to implement this.

jameslamb reviewed (changes requested):
Thanks very much for all of your efforts on this large feature. I believe all my previous requests have been addressed.

I gave this another review tonight and identified a few more things that I believe should be changed before this is merged.

cyfdecyf commented:
@jameslamb Thanks for those detailed review. I've applied you suggestions and made a few small tweaks for other parts of this PR.

jameslamb reviewed (approved):
I pulled this branch again today and rendered the documentation. All look ok to me, so I think all of my comments have been addressed.

Thanks very much!

StrikerRUS commented on include/LightGBM/c_api.h:
```diff
@@ -214,6 +214,17 @@ LIGHTGBM_C_EXPORT int LGBM_DatasetCreateFromCSC(const void* col_ptr,
+ * \param[out] out Created indices, type is int32_t, caller should insure out contains enough space to hold indices
+ * \return 0 when succeed, -1 when failure happens
+ */
+LIGHTGBM_C_EXPORT int LGBM_SampleIndices(int32_t total_nrow,
```
I think it will be useful to return actual number of sampled rows (`int32_t* out_len`) as well as indices themselves. Caller can call this function first time to get number of elements, allocate memory, and call this function second time to get indices. Something similar to allocating string buffers.
https://github.com/microsoft/LightGBM/blob/c738c83bbddc80808b47d0603e182e9595b7b2f7/python-package/lightgbm/basic.py#L1878-L1900

Or we can define a separate helper function for calculating number of indices given particular config params.

I'm suggesting this because I don't like that we have multiple places of the following duplicated code:
```
int sample_cnt = static_cast<int>(total_nrow < config.bin_construct_sample_cnt ? total_nrow : config.bin_construct_sample_cnt);
```
at both Python and cpp sides.

cyfdecyf replied:
The key problem here is that we have to implement some logic in Python and make the behavior the same with cpp side.

Feature request #4321 can help sync default values in Python and cpp side. But it's not feasible to completely remove logic duplication in both side. In this PR, the sampling data from multiple file logic can't be easily implemented in cpp side.

I'm going to create a function to get sample count in Python side to remove duplication. For cpp side, I separate PR is more appropriate.

cyfdecyf replied:
Commit 96ebbfe7 adds a separate function `_get_sample_count` in `basic.py` to avoid code duplicate in Python side.

StrikerRUS replied:
Thanks! But why we can't add a separate function **at cpp side** and call it from Python in `_get_sample_count()` function? So that computation of `sample_cnt` will be performed only in one place - at cpp side.

StrikerRUS replied:
Then we'll be able to do something similar to
https://github.com/microsoft/LightGBM/blob/08d1ce4bbf9391f5cc945f5f015840f4a57774a6/include/LightGBM/c_api.h#L637

cyfdecyf replied:
Just looked at `c_api.cpp` and I've got a few questions:

1. Is it find to use one API (e.g. `LGBM_SampleCount`) inside other APIs? This is to remove code duplicate in cpp side.
2. If it's OK to the 1st question, let's add API `LGBM_SampleCount` first, then send a separate PR to remove code duplicate in cpp side?

This removes the need of syncing default sample count, but still leaves `ZERO_THRESHOLD`.

StrikerRUS replied:
1. TBH, I don't know. But we can create internal function and call it from both (`LGBM_SampleCount` and `LGBM_SampleIndices`) C API functions if it will be a problem.
2. OK, agree with you about the follow-up PR to not overcomplicate this PR and its' discussion.

> This removes the need of syncing default sample count, but still leaves `ZERO_THRESHOLD`.

OK, no problem! My suggestion was to remove duplicated **code**. For duplicated definitions of constants you've alredy created a separate issue - thanks a lot.

cyfdecyf replied:
C API `LGBM_SampleCount` added.

StrikerRUS reviewed (commented):
@cyfdecyf 
Finally got a chance to take a look at this PR. Please check some my comments below.

cyfdecyf commented on examples/python-guide/dataset_from_multi_hdf5.py:
```diff
@@ -0,0 +1,102 @@
+        data.append(HDFSequence(f['X'], batch_size))
+        ylist.append(f['Y'][:])
+
+    # params = {
+    #     'bin_construct_sample_cnt': 200000,
+    #     'max_bin': 255,
+    # }
+    params = None
+    y = np.concatenate(ylist)
+    dataset = lgb.Dataset(data, label=y, params=params)
```
I hope this can serve as an example. I'll uncomment `params`.

cyfdecyf commented on examples/python-guide/dataset_from_multi_hdf5.py:
```diff
@@ -0,0 +1,102 @@
+    # The reason is that DataFrame column names will be used in Dataset. For a DataFrame with Int64Index
+    # as columns, Dataset will use column names like ["0", "1", "2", ...]. While for numpy array, column names
+    # are using the default one assigned in C++ code (dataset_loader.cpp), like ["Column_0", "Column_1", ...].
+    dataset.save_binary('regression.train.from_hdf.bin')
```
Yes. `.gitignore` contains a line to ignore `*.bin`.

cyfdecyf commented on examples/python-guide/dataset_from_multi_hdf5.py:
```diff
@@ -0,0 +1,102 @@
+    create_dataset_from_multiple_hdf(
+        ['regression1.h5', 'regression2.h5'],
+        batch_size=batch_size,
+    )
```
Fixed in commit ebdada5d.

cyfdecyf commented:
@StrikerRUS Thanks for the review. I've addressed your comments and pushed my changes.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1262,8 +1482,15 @@ def _lazy_init(self, data, label=None, reference=None,
+        elif isinstance(data, list) and len(data) > 0:
+            if all(isinstance(x, np.ndarray) for x in data):
+                self.__init_from_list_np2d(data, params_str, ref_dataset)
+            elif all(Sequence.is_class(x) for x in data):
```
`isinstance()` returns `True` for subclasses. Is this what `is_class` was designed for?
```suggestion
            elif all(isinstance(x, Sequence) for x in data):
```

cyfdecyf replied:
This is to allow user to implement their own `Sequence` without subclassing. The only interface required now is `__len__` and `__getitem__`.

`Sequence` is an abstract class and uses `isinstance` here at first in this PR. We removed relying on `abc` and switched to this `Sequence.is_class` function later.

StrikerRUS replied:
Sorry, I'm not 100% sold on this.

For your example one can simply write something like the following:
```python
try:
    from lightgbm import Sequence as lgb_Sequence
except ImportError:
    class lgb_Sequence:
        pass

...


class MySequence(lgb_Sequence):
   ...
```
and a library will support LightGBM `Sequence` interface only if LightGBM is installed.

I strongly believe that first version "Sequence is an abstract class and uses `isinstance` here at first in this PR" was better design.

cyfdecyf replied:
I agree making `Sequence` an abstract class is better. The `Sequence.is_class` method is tricky to get right. I'll change `Sequence` back to an abstract class.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1262,8 +1482,15 @@ def _lazy_init(self, data, label=None, reference=None,
+                self.__init_from_seqs(data, ref_dataset)
+            else:
+                raise TypeError('Data list can only be of ndarray or Sequence')
+        elif Sequence.is_class(data):
```
```suggestion
        elif isinstance(data, Sequence):
```

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1291,6 +1518,75 @@ def _lazy_init(self, data, label=None, reference=None,
+                seq_id += 1
+                seq = seqs[seq_id]
+            id_in_seq = row_id - offset
+            row = seq[int(id_in_seq)]
```
Sorry, can get it how `id_in_seq` could be not int?
```suggestion
            row = seq[id_in_seq]
```

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1291,6 +1518,75 @@ def _lazy_init(self, data, label=None, reference=None,
+
+        return filtered, filtered_idx
+
+    def __init_from_seqs(self, seqs: List[Sequence], ref_dataset: 'Dataset'):
```
According to the check below
```
# create validation dataset from ref_dataset
if ref_dataset:
```
```suggestion
    def __init_from_seqs(self, seqs: List[Sequence], ref_dataset: Optional['Dataset'] = None):
```
(`Optional` should be imported.)

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1291,6 +1518,75 @@ def _lazy_init(self, data, label=None, reference=None,
+        total_nrow = sum(len(seq) for seq in seqs)
+
+        # create validation dataset from ref_dataset
+        if ref_dataset:
```
I believe more explicit check is better.
```suggestion
        if ref_dataset is not None:
```

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1291,6 +1518,75 @@ def _lazy_init(self, data, label=None, reference=None,
+            batch_size = seq.batch_size or Sequence.batch_size
+            for start in range(0, nrow, batch_size):
+                end = min(start + batch_size, nrow)
+                self.push_rows(seq[start:end])
```
```suggestion
                self.push_rows(seq[start:end])
        return self
```

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1188,156 @@ def __init__(self, data, label=None, reference=None,
+        indices : numpy array
+            Indices for sampled data.
+        """
+        param_str = param_dict_to_str(self.params)
```
This is filtered version of all params. Also, benefits from checking param aliases.
https://github.com/microsoft/LightGBM/blob/4167649cd50035b8237d121f706dc9f1e1b79ed0/python-package/lightgbm/basic.py#L1121-L1128

```suggestion
        param_str = param_dict_to_str(self.get_params())
```

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1188,156 @@ def __init__(self, data, label=None, reference=None,
+        """
+        param_str = param_dict_to_str(self.params)
+        # Note self.params may contain 'bin_construct_sample_cnt' but is None.
+        sample_cnt = _get_sample_count(self.params, total_nrow)
```
```suggestion
        sample_cnt = _get_sample_count(self.get_params(), total_nrow)
```

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1188,156 @@ def __init__(self, data, label=None, reference=None,
+        num_per_col_ptr, _, _ = c_int_array(num_per_col)
+
+        self.handle = ctypes.c_void_p()
+        params_str = param_dict_to_str(self.params)
```
```suggestion
        params_str = param_dict_to_str(self.get_params())
```

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1291,6 +1518,75 @@ def _lazy_init(self, data, label=None, reference=None,
+        if ref_dataset:
+            self.init_from_ref_dataset(total_nrow, ref_dataset)
+        else:
+            sample_cnt = _get_sample_count(self.params, total_nrow)
```
```suggestion
            sample_cnt = _get_sample_count(self.get_params(), total_nrow)
```

StrikerRUS reviewed (commented):
@cyfdecyf Thanks a lot for addressing my previous batch of comments. I commented about cpp function for getting samples_cnt above.

Here are some new comments..

cyfdecyf commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1188,156 @@ def __init__(self, data, label=None, reference=None,
+            Constructed Dataset object.
+        """
+        ncol = len(sample_indices)
+        assert len(sample_data) == ncol, "#sample data column != #column indices"
```
Sampling is done internally at LightGBM's python-package, thus I think this internal checking and marks a BUG if triggered. (`init_from_sample` is private.)

StrikerRUS replied:
OK, got it! Agree with you.

cyfdecyf commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1188,156 @@ def __init__(self, data, label=None, reference=None,
         except AttributeError:
             pass
 
+    def create_sample_indices(self, total_nrow: int) -> np.ndarray:
```
Yes, this is private.

To keep consistency, I used double underscore to mark private functions in this PR.

Should we keep using double underscore or just use single one to mark private from now on?

StrikerRUS replied:
I personally don't support using double underscore without serious reasons, but I'm totally fine with the way you went.

> There's a concept in Python: "we're all consenting adults here". If you use the `__private` form, who are you protecting the attribute from? It's the responsibility of subclasses to use attributes from superclasses properly, and it's the responsibility of superclasses to document their attributes properly.
It's better to use the single-leading-underscore convention, `_internal`. "This isn't name mangled at all; it just indicates to others to "be careful with this, it's an internal implementation detail; don't touch it if you don't fully understand it". It's only a convention though.
   https://stackoverflow.com/a/6930223

cyfdecyf commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1188,156 @@ def __init__(self, data, label=None, reference=None,
+        ))
+        return self
+
+    def push_rows(self, data: np.ndarray) -> 'Dataset':
```
For this PR this is private.

We have an internal branch where push_rows is public and is called by `Dataset` user. This allows us to create multiple dataset binaries with different parameters in a single pass to the entire input data.

cyfdecyf commented on python-package/lightgbm/basic.py:
```diff
@@ -1111,13 +1188,156 @@ def __init__(self, data, label=None, reference=None,
+        ))
+        return self
+
+    def init_from_sample(
```
Our internal branch exposes all sampling related interfaces. For this PR we can make this private.

StrikerRUS replied:
Thanks a lot for the explanation! I believe for now it's enough to accept `Sequence` in constructor and keep all machinery private. Advanced users can check the source code and use internal methods if really needed.

cyfdecyf commented on python-package/lightgbm/__init__.py:
```diff
@@ -29,7 +29,7 @@
         __version__ = version_file.read().strip()
 
-__all__ = ['Dataset', 'Booster', 'CVBooster',
+__all__ = ['Dataset', 'Booster', 'CVBooster', 'Sequence',
```
I'll do this after commiting your suggestions.

StrikerRUS commented on docs/Python-API.rst:
```diff
@@ -12,6 +12,7 @@ Data Structure API
     Dataset
     Booster
     CVBooster
+    Sequence
```
Please mention `Sequence` here:
https://github.com/microsoft/LightGBM/blob/master/docs/Python-Intro.rst#data-interface

StrikerRUS commented on docs/Python-API.rst:
```diff
@@ -12,6 +12,7 @@ Data Structure API
     Booster
     CVBooster
+    Sequence
 
```
Please add `Sequence` here
https://github.com/microsoft/LightGBM/blob/c7134fa7ccbe1a1b1819ac2a66407504d3fab27a/python-package/lightgbm/basic.py#L1058

StrikerRUS commented on docs/Python-API.rst:
```diff
@@ -12,6 +12,7 @@ Data Structure API
     CVBooster
+    Sequence
 
 Training API
```
Please add `Sequence` here
https://github.com/microsoft/LightGBM/blob/c7134fa7ccbe1a1b1819ac2a66407504d3fab27a/python-package/lightgbm/basic.py#L1944

cyfdecyf replied:
Let's add this in follow-up PR.

StrikerRUS commented on docs/Python-API.rst:
```diff
@@ -12,6 +12,7 @@ Data Structure API
+    Sequence
 
 Training API
 ------------
```
Please consider updating the following two methods:
https://github.com/microsoft/LightGBM/blob/c7134fa7ccbe1a1b1819ac2a66407504d3fab27a/python-package/lightgbm/basic.py#L1939
https://github.com/microsoft/LightGBM/blob/c7134fa7ccbe1a1b1819ac2a66407504d3fab27a/python-package/lightgbm/basic.py#L2048

cyfdecyf replied:
I've never used these two APIs. I need some time to figure out how to change these two methods.

StrikerRUS replied:
OK. I think it's totally fine to do this in a follow-up PR. I'll create an issue referring this comment to not lose our discussion after merging this PR.

StrikerRUS replied:
Created #4403 to track this enhancement.

cyfdecyf replied:
Thanks for the separate issue.

StrikerRUS reviewed (commented):
We need some more docs updates. Also, please check whether we can do something smart with subsetting in `get_data()` and in `add_features_from()` methods of `Dataset` class when `seld.data` (and/or `other.data`) is `Sequence`.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -591,6 +600,77 @@ def _load_pandas_categorical(file_name=None, model_str=None):
+        >>> seq.batch_size
+
+    - With random access, **data sampling does not need to go through all data**.
+    - With range data access, there's **no need to read all data into memory thus reduce memory usage**.
```
Document `batch_size`.
```suggestion
    - With range data access, there's **no need to read all data into memory thus reduce memory usage**.

    Attributes
    ----------
    batch_size : int
        Default size of a batch.
```

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -89,6 +91,115 @@ def test_basic(tmp_path):
+
+    assert filecmp.cmp(npy_bin_fname, seq_bin_fname)
+
+    # Test for validation set.
```
Sorry, can get it. Is validation dataset creation supported (#4184)?
If so, please add `Sequence` here 
https://github.com/microsoft/LightGBM/blob/c7134fa7ccbe1a1b1819ac2a66407504d3fab27a/python-package/lightgbm/basic.py#L1477

cyfdecyf replied:
Yes. We can create valid dataset from `Sequence`. I'll update doc.

StrikerRUS replied:
Was #4184 created for discussion purposes or that issue is for some other feature request? In other words, can we close it after merging this PR?

cyfdecyf replied:
It's more for ciscussion purpose.

cyfdecyf commented on tests/python_package_test/test_basic.py:
```diff
@@ -89,6 +91,115 @@ def test_basic(tmp_path):
+        else:
+            raise TypeError(f"Sequence Index must be an integer/list/slice, got {type(idx)}")
+
+    def _get_one_line(self, idx):
+        return self.ndarray[idx]
```
The purpose here is to serve as an example and make it more clear what the implementation should do.

I'm going to remove this because this should be very easy to understand even without this function.

cyfdecyf commented on tests/python_package_test/test_basic.py:
```diff
@@ -89,6 +91,115 @@ def test_basic(tmp_path):
+    return seqs
+
+
+@pytest.mark.parametrize('sample_count', [11, 23, 100, None])
+@pytest.mark.parametrize('batch_size', [3, 20, None])
+@pytest.mark.parametrize('include_0', [False, True])
+@pytest.mark.parametrize('include_nan', [False, True])
+@pytest.mark.parametrize('num_seq', [1, 3])
```
I'll reduce number of test according to your suggestions.

cyfdecyf commented:
@StrikerRUS Thanks for your comments.

While adding C API `LGBM_SampleCount`, I noticed `rand.Sample` may return number of values less than requested sample count.

https://github.com/microsoft/LightGBM/blob/0ca2c49424ffc93bee6b70a44c874e0db89cb490/include/LightGBM/utils/random.h#L63-L101

Guess this is why `c_api.cpp` has code like following

https://github.com/microsoft/LightGBM/blob/0ca2c49424ffc93bee6b70a44c874e0db89cb490/src/c_api.cpp#L1067-L1070

If my understanding to `Random::Sample` is correct, then I suggest update it's API doc.

After this PR is merged, the above sample indices logic can be replaced as:

```
auto sample_indices = CreateSampleIndices(total_nrow, config);
sample_cnt = static_cast<int>(sample_indices.size());
```

StrikerRUS commented on include/LightGBM/c_api.h:
```diff
@@ -53,6 +53,30 @@ LIGHTGBM_C_EXPORT const char* LGBM_GetLastError();
+                                       int* out);
+
+/*!
+ * \brief Create sample indices for total number of rows.
+ * \param total_nrow Number of all data rows
+ * \param parameters Additional parameters, specify sample count and random seed in parameter
+ * \param[out] out Created indices, type is int32_t, caller should insure out contains enough space to hold indices
+ * \param[out_len] out Number of indices. This maybe less than the one returned by ``LGBM_SampleCount``.
+ * \return 0 when succeed, -1 when failure happens
```
```suggestion
 * \brief Create sample indices for total number of rows.
 * \note
 * You should pre-allocate memory for ``out``, you can get its length by ``LGBM_SampleCount``.
 * \param total_nrow Number of all data rows
 * \param parameters Additional parameters, namely, ``bin_construct_sample_cnt`` and ``data_random_seed`` are used to produce the output
 * \param[out] out Created indices, type is int32_t
 * \param[out] out_len Number of indices. This maybe less than the one returned by ``LGBM_SampleCount``
 * \return 0 when succeed, -1 when failure happens
```

StrikerRUS commented on include/LightGBM/c_api.h:
```diff
@@ -53,6 +53,30 @@ LIGHTGBM_C_EXPORT const char* LGBM_GetLastError();
+ * \param[out_len] out Number of indices. This maybe less than the one returned by ``LGBM_SampleCount``.
+ * \return 0 when succeed, -1 when failure happens
+ */
+LIGHTGBM_C_EXPORT int LGBM_SampleIndices(int32_t total_nrow,
+                                         const char* parameters,
+                                         void* out,
+                                         int64_t* out_len);
```
I find it confusing that ``out_len`` type is `int64_t` while type of `total_nrow` is `int32_t`. I believe number of sampled rows cannot be bigger than total number of rows.

cyfdecyf replied:
It's better to change `out_len` to `int32_t`. Done in commit a5bb2e6 and fe049f2.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -89,6 +91,103 @@ def test_basic(tmp_path):
+    valid_ds.save_binary(valid_npy_bin_fname)
+
+    valid_seqs = _create_sequence_from_ndarray(valid_X, num_seq, batch_size)
+    valid_seq_ds = lgb.Dataset(valid_seqs, label=valid_Y, params=params, reference=valid_ds)
```
Please add one more case when `Dataset` is created from sequences and then validation `Dataset` is also created from sequences via `create_valid()` method of the first Datset.

cyfdecyf replied:
Done in a3a9c4a.

In the original code in the comment, `reference` should be set to original `Dataset` instead of another validation dataset. This is also fixed to use dataset from numpy array.

StrikerRUS replied:
Ah, nice catch! Thanks very much!

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1,5 +1,6 @@
 # coding: utf-8
```
Please add `Sequence` here
https://github.com/microsoft/LightGBM/blob/c7134fa7ccbe1a1b1819ac2a66407504d3fab27a/python-package/lightgbm/basic.py#L1944

cyfdecyf replied:
Since we are going to support `Sequence` in follow-up PR, I'm going to mark this as resoved for now.

StrikerRUS commented on docs/Python-API.rst:
```diff
@@ -12,6 +12,7 @@ Data Structure API
     Dataset
```
Please mention `Sequence` here:
https://github.com/microsoft/LightGBM/blob/master/docs/Python-Intro.rst#data-interface

cyfdecyf replied:
I've add some doc. Please take a look to see if this is appropriate.

StrikerRUS reviewed (commented):
Another round of review comments from me which includes some unresolved comments from previous ones to not ask you re-read all previous comments.

StrikerRUS commented:
> While adding C API `LGBM_SampleCount`, I noticed `rand.Sample` may return number of values less than requested sample count.
   https://github.com/microsoft/LightGBM/pull/4089#issuecomment-865455814

@shiyu1994 Could you please take a look?

cyfdecyf commented:
@StrikerRUS Thanks again for your review. I've pushed my changes to address all the comments.

shiyu1994 commented:
> > While adding C API `LGBM_SampleCount`, I noticed `rand.Sample` may return number of values less than requested sample count.
> > [#4089 (comment)](https://github.com/microsoft/LightGBM/pull/4089#issuecomment-865455814)
> 
> @shiyu1994 Could you please take a look?

I think `Sample` only returns `K` elements in expectation, but does not guarantee to return exactly `K` elements. We need to figure out an efficient way to `Sample` exactly `K` elements.

StrikerRUS commented on docs/Python-Intro.rst:
```diff
@@ -69,6 +71,38 @@ Many of the examples in this page use functionality from ``numpy``. To run the e
+- Reading data in batch thus saves memory when constructing ``Dataset`` object
+- Supports creating ``Dataset`` from multiple data files
+
+Please refer to ``Sequence`` `API doc <https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.CVBooster.html#lightgbm.Sequence>`__.
+`dataset_from_multi_hdf5.py <https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/dataset_from_multi_hdf5.py>`__
+is a detailed example.
```
API doc link will be version-independent (currently it's hardcoded to the `latest` version).

It's hard to spot where one link ends and another one starts, so I think it's better to split lines.
```suggestion
Please refer to ``Sequence`` `API doc <./Python-API.rst#data-structure-api>`__.

`dataset_from_multi_hdf5.py <https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/dataset_from_multi_hdf5.py>`__ is a detailed example.
```

StrikerRUS commented on include/LightGBM/c_api.h:
```diff
@@ -53,6 +53,32 @@ LIGHTGBM_C_EXPORT const char* LGBM_GetLastError();
+/*!
+ * \brief Create sample indices for total number of rows.
+ * \note
+ * You should pre-allocate memory for ``out``, you can get its length by ``LGBM_SampleCount``.
+ * \param num_total_row Number of total rows
+ * \param parameters Additional parameters, namely, ``bin_construct_sample_cnt`` and ``data_random_seed`` are used to produce the output
+ * \param[out] out Created indices, type is int32_t
+ * \param[out] out_len Number of indices. This maybe less than the one returned by ``LGBM_SampleCount``
```
```suggestion
 * You should pre-allocate memory for ``out``, you can get its length by ``LGBM_GetSampleCount``.
 * \param num_total_row Number of total rows
 * \param parameters Additional parameters, namely, ``bin_construct_sample_cnt`` and ``data_random_seed`` are used to produce the output
 * \param[out] out Created indices, type is int32_t
 * \param[out] out_len Number of indices. This maybe less than the one returned by ``LGBM_GetSampleCount``
```

StrikerRUS commented on src/c_api.cpp:
```diff
@@ -894,6 +894,51 @@ int LGBM_RegisterLogCallback(void (*callback)(const char*)) {
+                     int* out) {
+  API_BEGIN();
+  if (out == nullptr) {
+    Log::Fatal("LGBM_SampleCount output is nullptr");
```
```suggestion
    Log::Fatal("LGBM_GetSampleCount output is nullptr");
```

StrikerRUS commented on include/LightGBM/c_api.h:
```diff
@@ -53,6 +53,32 @@ LIGHTGBM_C_EXPORT const char* LGBM_GetLastError();
+ * \param[out] out Number of samples. This value is used to pre-allocate memory to hold sample indices when calling ``LGBM_SampleIndices``
+ * \return 0 when succeed, -1 when failure happens
+ */
+LIGHTGBM_C_EXPORT int LGBM_GetSampleCount(int32_t num_total_row,
+                                       const char* parameters,
+                                       int* out);
```
```suggestion
LIGHTGBM_C_EXPORT int LGBM_GetSampleCount(int32_t num_total_row,
                                          const char* parameters,
                                          int* out);
```

StrikerRUS commented on src/c_api.cpp:
```diff
@@ -894,6 +894,51 @@ int LGBM_RegisterLogCallback(void (*callback)(const char*)) {
+  return rand.Sample(total_nrow, sample_cnt);
+}
+
+int LGBM_GetSampleCount(int32_t num_total_row,
+                     const char* parameters,
+                     int* out) {
```
```suggestion
int LGBM_GetSampleCount(int32_t num_total_row,
                        const char* parameters,
                        int* out) {
```

StrikerRUS reviewed (commented):
Just some minor cleaning-up comments below.

StrikerRUS commented:
@shiyu1994 
> We need to figure out an efficient way to `Sample` exactly `K` elements.

Thanks a lot for taking a look! Do you think we need feature request for this?

Also, please update your review for this PR because there have been a lot of code changes at cpp side since your latest approval (now there are 2 new C API functions).

cyfdecyf commented:
@StrikerRUS Thanks for the cleanup. I should have done renaming with find and replace..

shiyu1994 commented:
@StrikerRUS Write to #2302. Sure, I'll update my review now.

shiyu1994 reviewed (commented):
I've checked again, only a few typos to be fixed and `DEFAULT_BIN_CONSTRUCT_SAMPLE_CNT` to be removed.

cyfdecyf commented on python-package/lightgbm/basic.py:
```diff
@@ -9,14 +10,27 @@
 from .libpath import find_lib_path
 
+ZERO_THRESHOLD = 1e-35
+DEFAULT_BIN_CONSTRUCT_SAMPLE_CNT = 200000
```
Yes, this can be removed now.

cyfdecyf commented:
@shiyu1994 Thanks for the review. Just addressed all the comments and pushed my changes.

shiyu1994 commented:
@cyfdecyf @StrikerRUS Sorry, after a careful investigation, I found that the `Sample` function returns exactly `K` elements. I should delete the feature request. But I don't know why the `sample_cnt` should be recalculated as mentioned here https://github.com/microsoft/LightGBM/pull/4089#issuecomment-865455814.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -591,6 +604,65 @@ def _load_pandas_categorical(file_name=None, model_str=None):
+    """
+    Generic data access interface.
+
+    Object should support the following operations:
+
+        # Get total row number.
```
I'm not sure whether double `:` was a typo or not (https://github.com/microsoft/LightGBM/pull/4089#discussion_r660289485). But it was acting as code block directive in Sphinx. Now without it we have the following:

![image](https://user-images.githubusercontent.com/25141164/123960871-7b410280-d9b8-11eb-950c-998c61cdd224.png)


I believe that explicitly mark this description as code block will provide a better rendered result:

![image](https://user-images.githubusercontent.com/25141164/123961060-a3c8fc80-d9b8-11eb-981c-da7c59ae67be.png)

cc @shiyu1994 

```suggestion
    Object should support the following operations:

    .. code-block::

        # Get total row number.
```

shiyu1994 replied:
Sorry, my bad. I think your modification is better.

cyfdecyf replied:
I actually tried single `:` and run `make html` but found no change. Guess it's browser cache not refreshing.

StrikerRUS reviewed (approved):
Thank you so so much for this awesome contribution!
LGTM! I don't have any other comments. Thanks for your patience!

cyfdecyf commented:
Commit 8a0f07ae4b836cb1b4e0725597f387afbdac9b1d introduced `sample_cnt` assignement from sampled indices vector.

In line 87 in the following code, if `NextShort` always return a value near `max_step`, then we may sample less then `K` values.

https://github.com/microsoft/LightGBM/blob/8a0f07ae4b836cb1b4e0725597f387afbdac9b1d/include/LightGBM/utils/random.h#L82-L91

shiyu1994 commented:
I see. But it seems that the latest version of `Sample` won't have such problem.
https://github.com/microsoft/LightGBM/blob/6a195a1bec42996b93eba02a511541a5615d068c/include/LightGBM/utils/random.h#L88-L95
In the loop above, each iteration a new element will be added into the set. But as #4371 pointed out, there's a bug in line 91, `NextInt(0, r)` should be `NextInt(0, r + 1)`. I think we can fix this, and remove the recalculation of `sample_cnt` in the code in another PR.

shiyu1994 commented:
@StrikerRUS @jameslamb Thank you both for your careful review. Can we merge this PR now?

StrikerRUS commented:
Yeah, I believe we can continue (#4403 and https://github.com/microsoft/LightGBM/pull/4089#issuecomment-871958002) in a follow-up PRs.

@cyfdecyf Great contribution, many thanks!

cyfdecyf commented:
Thanks for all the reviews. I'll start working on #4403 next week.

StrikerRUS commented:
> I'll start working on #4403 next week.

@cyfdecyf
Many thanks!

Also, #4450 has been merged, so please check https://github.com/microsoft/LightGBM/pull/4089#issuecomment-871958002
> ... and remove the recalculation of `sample_cnt` in the code in another PR.
